### PR TITLE
File.rmdir!/1: give a less confusing error message when directory is not empty

### PR DESCRIPTION
**Pull request by eksperimental**
Closes https://github.com/elixir-lang/elixir/issues/5296

### CODE REVIEW

File: lib/elixir/lib/exception.ex [hunk 1]
```diff
@@ -787,7 +787,14 @@ defmodule File.Error do
   defexception [:reason, :path, action: ""]
 
   def message(exception) do
-    formatted = IO.iodata_to_binary(:file.format_error(exception.reason))
+    formatted =
+      case exception.reason do
+        :eexist ->
+          "directory is not empty"
```
Comment (josevalim) [inline lib/elixir/lib/exception.ex]: We should change this exclusively inside `rmdir!` as the "eexist" reason is valid in other situations. For example:

```iex
iex> File.mkdir "foo"
:ok
iex> File.mkdir "foo"
{:error, :eexist}
```

File: lib/elixir/test/elixir/file_test.exs [hunk 2]
```diff
@@ -1074,6 +1074,26 @@ defmodule FileTest do
       end
     end
 
+    test "rmdir! error messages" do
+      fixture = tmp_path("tmp_test")
+      File.mkdir_p(fixture)
+      File.touch (fixture <> "/file")
```
Comment (josevalim) [inline lib/elixir/test/elixir/file_test.exs]: Please remove space.

File: lib/elixir/test/elixir/file_test.exs [hunk 3]
```diff
@@ -1074,6 +1074,26 @@ defmodule FileTest do
...
+        File.rmdir!(fixture)
+      end
+
+      # directory does not exist
+      non_existent_dir = fixture <> "/non_existent_dir"
+      assert_raise File.Error, ~r"\Acould not remove directory #{inspect non_existent_dir}: (not a directory|no such file or directory)", fn ->
+        File.rmdir!(non_existent_dir)
+      end
+
+      assert {:ok, _} = File.rm_rf(fixture)
```
Comment (josevalim) [inline lib/elixir/test/elixir/file_test.exs]: Do we need this assertion? Is it part of our test?

File: lib/elixir/lib/exception.ex (hunk 1 above)
Comment (eksperimental) [inline lib/elixir/lib/exception.ex]: done!

File: lib/elixir/test/elixir/file_test.exs (hunk 2 above)
Comment (eksperimental) [inline lib/elixir/test/elixir/file_test.exs]: done

**Comment by eksperimental** [comment]
we are good to go

File: lib/elixir/lib/exception.ex [hunk 4]
```diff
@@ -787,7 +787,14 @@ defmodule File.Error do
   defexception [:reason, :path, action: ""]
 
   def message(exception) do
-    formatted = IO.iodata_to_binary(:file.format_error(exception.reason))
+    formatted =
+      case {exception.action, exception.reason} do
```
Comment (lexmag) [inline lib/elixir/lib/exception.ex]: We are accessing `exception` fields multiple times, what if we decompose exception on first line:
```elixir
%{action: action, reason: reason, path: path} = exception
formatted =
  case {action, reason} do
```
?

Comment (eksperimental) [inline lib/elixir/lib/exception.ex]: great suggestion.
I didn't create the exception variable though, as I see no use for it.

File: lib/elixir/test/elixir/file_test.exs [hunk 5]
```diff
@@ -1074,6 +1074,25 @@ defmodule FileTest do
       end
     end
 
+    test "rmdir! error messages" do
+      fixture = tmp_path("tmp_test")
+      File.mkdir_p(fixture)
+      File.touch(fixture <> "/file")
+
+      # directory is not empty
+      assert_raise File.Error, ~r"\Acould not remove directory #{inspect fixture}: directory is not empty", fn ->
```
Comment (whatyouhide) [inline lib/elixir/test/elixir/file_test.exs]: Do we need the regex here?

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Pull request merged**